keps/sig-node/4939-tls-in-grpc-probe/README.md

```diff
@@ -0,0 +1,735 @@
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+Plan is to add unit tests to `pkg/probe/grpc/grpc_test.go`
+that can verify the flags are interpreted correctly
+and that a TLS (with no verify) and non-TLS configurations work as expected.
+
+- `k8s.io/kubernetes/pkg/probe/grpc`: TBD
+
+##### Integration tests
+
+N/A, only unit tests and e2e coverage.
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+Tests in test/e2e/common/node/container_probe.go:
+- should not be restarted with a GRPC liveness probe: TBD
+- should be restarted with a GRPC liveness probe: TBD
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implement the feature.
+- Add unit and e2e tests for the feature.
+
+#### Beta
+
+- Solicit feedback from the Alpha.
+- Ensure tests are stable and passing.
+
+#### GA
+- [ ] Address feedback from beta usage
+- [ ] Validate that API is appropriate for users
+- [ ] Close on any remaining open issues & bugs
+
+### Upgrade / Downgrade Strategy
+
+Upgrade: default values for new configurables should default to the current state so upgrade should not require anything.
+
+Downgrade: gRPC probes will only support TLS in a downgrade from alpha
+
+### Version Skew Strategy
+
+N/A
```

Comment:
What happens when you have a back-rev kubelet?  It's allowed for nodes to be 3 releases behind the control plane, so what happens when apiserver accepts a pod spec with tls turned on and the pod lands on an old node where kubelet doesn't even know about TLS?  Your pod will run, but fail to probe.  This section is to call that out.

Then you get to discuss whether that's actually OK.

I don't know what the norm is for features which are entirely kubelet implemented.

* sit in alpha for 3 releases
* alpha for 1 release, beta+off for 2 more, then beta+on or GA
* alpha for 1 release, beta+on and let users figure out why backrev kubelets are a bad idea?

@tallclair @SergeyKanzhelev

Reply:
ah ok that makes sense, just updated the section to reflect that concern. will wait on guidance for the exact strategy.